File: hashmap/HashmapTest.hs
import Hashmap as HM
import Data.Maybe (isNothing)
import System.CPUTime
import Text.Printf
import Control.Monad (forM_, replicateM)
import Data.Functor ((<&>))
import Data.Maybe (isNothing, isJust)
import System.Random (randomRIO)
import Data.List (nub)
import Control.Exception (catch, SomeException)
import Data.List (foldl')

-- Timing utility
timeIt :: String -> IO a -> IO a
timeIt label action = do
    start <- getCPUTime
    result <- action
    end <- getCPUTime
    let diff = (fromIntegral (end - start) / (10^12)) :: Double
    printf "%s: %0.3f sec\n" label diff
    return result

-- Random string generator
randomString :: Int -> IO String
randomString len = do
    let chars = ['a'..'z'] ++ ['A'..'Z'] ++ ['0'..'9']
    sequence $ replicate len $ randomRIO (0, length chars - 1) >>= return . (chars !!)

-- Modify performance test functions for better results
testInsertPerformance :: IO ()
testInsertPerformance = do
    putStrLn "\nTesting Insert Performance:"
    forM_ [1000, 10000, 100000] $ \size -> do
        keys <- replicateM size (randomString 10)
        values <- replicateM size (randomString 10)
        let pairs = zip keys values
        timeIt (show size ++ " inserts") $ do
            let !final = foldl' (\hm (k,v) -> HM.insert k v hm) HM.empty pairs
            putStr $ "Size: " ++ show (HM.size final) ++ " "
            return ()

testLookupPerformance :: IO ()
testLookupPerformance = do
    putStrLn "\nTesting Lookup Performance:"
    -- Create smaller test data to avoid memory issues
    keys <- replicateM 10000 (randomString 10)
    values <- replicateM 10000 (randomString 10)
    let !hm = foldl' (\h (k,v) -> HM.insert k v h) HM.empty (zip keys values)
    
    forM_ [1000, 5000, 10000] $ \lookups -> do
        timeIt (show lookups ++ " lookups") $ do
            let !results = length [() | k <- take lookups (cycle keys), isJust (HM.lookup k hm)]
            putStr $ "Found: " ++ show results ++ " "
            return ()

testBulkOperations :: IO ()
testBulkOperations = do
    putStrLn "\nTesting Bulk Operations Performance:"
    forM_ [1000, 5000, 10000] $ \size -> do
        keys <- replicateM size (randomString 10)
        values <- replicateM size (randomString 10)
        let pairs = zip keys values
        putStrLn $ "\nSize " ++ show size ++ ":"
        timeIt "fromList" $ do
            let !hm = HM.fromList pairs
            putStr $ "Size: " ++ show (HM.size hm) ++ " "
            return ()
        timeIt "toList" $ do
            let !hm = HM.fromList pairs
            let !lst = HM.toList hm
            putStr $ "Length: " ++ show (length lst) ++ " "
            return ()

-- Test cases for the HashMap
testInsert :: IO ()
testInsert = do
    let hm = HM.empty
    let hm' = HM.insert "key1" "value1" hm
    if HM.lookup "key1" hm' == Just "value1" && isNothing (HM.lookup "key1" hm)
        then putStrLn "testInsert passed"
        else putStrLn "testInsert failed"

testLookup :: IO ()
testLookup = do
    let hm = HM.insert "key1" "value1" $ HM.insert "key2" "value2" HM.empty
    if HM.lookup "key1" hm == Just "value1" && HM.lookup "key2" hm == Just "value2" && isNothing (HM.lookup "key3" hm)
        then putStrLn "testLookup passed"
        else putStrLn "testLookup failed"

testDelete :: IO ()
testDelete = do
    let hm = HM.fromList [("key1", "value1"), ("key2", "value2")]
    let hm' = HM.delete "key1" hm
    if isNothing (HM.lookup "key1" hm') && HM.lookup "key2" hm' == Just "value2" && HM.lookup "key1" hm == Just "value1"
        then putStrLn "testDelete passed"
        else putStrLn "testDelete failed"

testFromList :: IO ()
testFromList = do
    let hm = HM.fromList [("key1", "value1"), ("key2", "value2")]
    if HM.lookup "key1" hm == Just "value1" && HM.lookup "key2" hm == Just "value2"
        then putStrLn "testFromList passed"
        else putStrLn "testFromList failed"

testToList :: IO ()
testToList = do
    let hm = HM.fromList [("key1", "value1"), ("key2", "value2")]
    if HM.toList hm == [("key1", "value1"), ("key2", "value2")]
        then putStrLn "testToList passed"
        else putStrLn "testToList failed"

-- Additional tests to verify purely functional behavior
testPurelyFunctionalInsert :: IO ()
testPurelyFunctionalInsert = do
    let hm = HM.empty
    let hm' = HM.insert "key1" "value1" hm
    let hm'' = HM.insert "key2" "value2" hm
    if HM.lookup "key1" hm' == Just "value1" && isNothing (HM.lookup "key1" hm) && isNothing (HM.lookup "key2" hm)
        then putStrLn "testPurelyFunctionalInsert passed"
        else putStrLn "testPurelyFunctionalInsert failed"

testPurelyFunctionalDelete :: IO ()
testPurelyFunctionalDelete = do
    let hm = HM.fromList [("key1", "value1"), ("key2", "value2")]
    let hm' = HM.delete "key1" hm
    let hm'' = HM.delete "key2" hm
    if isNothing (HM.lookup "key1" hm') && HM.lookup "key2" hm' == Just "value2" && HM.lookup "key1" hm == Just "value1"
        then putStrLn "testPurelyFunctionalDelete passed"
        else putStrLn "testPurelyFunctionalDelete failed"

-- Tests for new operations
testUpdate :: IO ()
testUpdate = do
    let hm = HM.insert "key1" "value1" HM.empty
    let hm' = HM.update "key1" "newValue1" hm
    if HM.lookup "key1" hm' == Just "newValue1" && HM.lookup "key1" hm == Just "value1"
        then putStrLn "testUpdate passed"
        else putStrLn "testUpdate failed"

testMerge :: IO ()
testMerge = do
    let hm1 = HM.fromList [("key1", "value1"), ("key2", "value2")]
    let hm2 = HM.fromList [("key2", "newValue2"), ("key3", "value3")]
    let hm' = HM.merge hm1 hm2
    if HM.lookup "key1" hm' == Just "value1" && HM.lookup "key2" hm' == Just "value2" && HM.lookup "key3" hm' == Just "value3"
        then putStrLn "testMerge passed"
        else putStrLn "testMerge failed"

testKeys :: IO ()
testKeys = do
    let hm = HM.fromList [("key1", "value1"), ("key2", "value2")]
    if HM.keys hm == ["key1", "key2"]
        then putStrLn "testKeys passed"
        else putStrLn "testKeys failed"

testValues :: IO ()
testValues = do
    let hm = HM.fromList [("key1", "value1"), ("key2", "value2")]
    if HM.values hm == ["value1", "value2"]
        then putStrLn "testValues passed"
        else putStrLn "testValues failed"

testSize :: IO ()
testSize = do
    let hm = HM.fromList [("key1", "value1"), ("key2", "value2")]
    if HM.size hm == 2
        then putStrLn "testSize passed"
        else putStrLn "testSize failed"

testMapValues :: IO ()
testMapValues = do
    let hm = HM.fromList [("key1", "value1"), ("key2", "value2")]
    let hm' = HM.mapValues (++ "_updated") hm
    if HM.lookup "key1" hm' == Just "value1_updated" && HM.lookup "key2" hm' == Just "value2_updated"
        then putStrLn "testMapValues passed"
        else putStrLn "testMapValues failed"

-- Run all tests
main :: IO ()
main = do
    putStrLn "Running functional tests:"
    testInsert
    testLookup
    testDelete
    testFromList
    testToList
    testPurelyFunctionalInsert
    testPurelyFunctionalDelete
    testUpdate
    testMerge
    testKeys
    testValues
    testSize
    testMapValues
    
    -- Modified performance tests with error handling
    catch testInsertPerformance (\e -> putStrLn $ "Insert test error: " ++ show (e :: SomeException))
    catch testLookupPerformance (\e -> putStrLn $ "Lookup test error: " ++ show (e :: SomeException))
    catch testBulkOperations (\e -> putStrLn $ "Bulk operations test error: " ++ show (e :: SomeException))

File: hashmap/HashmapTest.cpp
#include "Hashmap.cpp"
#include <iostream>
#include <chrono>
#include <random>
#include <sstream>

// Helper function for timing measurements
template<typename F>
double measureTime(F func) {
    auto start = std::chrono::high_resolution_clock::now();
    func();
    auto end = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> diff = end - start;
    return diff.count();
}

// Generate random string of given length
std::string randomString(int length) {
    static const char alphanum[] =
        "0123456789"
        "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
        "abcdefghijklmnopqrstuvwxyz";
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> dis(0, sizeof(alphanum) - 2);
    std::string str;
    str.reserve(length);
    for (int i = 0; i < length; ++i) {
        str += alphanum[dis(gen)];
    }
    return str;
}

void testInsert() {
    HashMap hm;
    hm.insert("key1", "value1");
    if (hm.lookup("key1") == "value1") {
        std::cout << "testInsert passed" << std::endl;
    } else {
        std::cout << "testInsert failed" << std::endl;
    }
}

void testLookup() {
    HashMap hm;
    hm.insert("key1", "value1");
    hm.insert("key2", "value2");
    if (hm.lookup("key1") == "value1" && hm.lookup("key2") == "value2" && hm.lookup("key3") == "") {
        std::cout << "testLookup passed" << std::endl;
    } else {
        std::cout << "testLookup failed" << std::endl;
    }
}

void testDelete() {
    HashMap hm;
    hm.insert("key1", "value1");
    hm.insert("key2", "value2");
    hm.deleteKey("key1");
    if (hm.lookup("key1") == "" && hm.lookup("key2") == "value2") {
        std::cout << "testDelete passed" << std::endl;
    } else {
        std::cout << "testDelete failed" << std::endl;
    }
}

void testFromList() {
    HashMap hm;
    hm.fromList({{"key1", "value1"}, {"key2", "value2"}});
    if (hm.lookup("key1") == "value1" && hm.lookup("key2") == "value2") {
        std::cout << "testFromList passed" << std::endl;
    } else {
        std::cout << "testFromList failed" << std::endl;
    }
}

void testToList() {
    HashMap hm;
    hm.fromList({{"key1", "value1"}, {"key2", "value2"}});
    auto list = hm.toList();
    std::unordered_map<std::string, std::string> expected = {{"key1", "value1"}, {"key2", "value2"}};
    bool passed = true;
    for (const auto& pair : list) {
        if (expected[pair.first] != pair.second) {
            passed = false;
            break;
        }
    }
    if (passed) {
        std::cout << "testToList passed" << std::endl;
    } else {
        std::cout << "testToList failed" << std::endl;
    }
}

// Additional tests to verify purely functional behavior
void testPurelyFunctionalInsert() {
    HashMap hm;
    hm.insert("key1", "value1");
    HashMap hm2 = hm;
    hm.insert("key2", "value2");
    if (hm.lookup("key1") == "value1" && hm.lookup("key2") == "value2" && hm2.lookup("key2") == "") {
        std::cout << "testPurelyFunctionalInsert passed" << std::endl;
    } else {
        std::cout << "testPurelyFunctionalInsert failed" << std::endl;
    }
}

void testPurelyFunctionalDelete() {
    HashMap hm;
    hm.insert("key1", "value1");
    hm.insert("key2", "value2");
    HashMap hm2 = hm;
    hm.deleteKey("key1");
    if (hm.lookup("key1") == "" && hm.lookup("key2") == "value2" && hm2.lookup("key1") == "value1") {
        std::cout << "testPurelyFunctionalDelete passed" << std::endl;
    } else {
        std::cout << "testPurelyFunctionalDelete failed" << std::endl;
    }
}

// Tests for new operations
void testUpdate() {
    HashMap hm;
    hm.insert("key1", "value1");
    hm.update("key1", "newValue1");
    if (hm.lookup("key1") == "newValue1") {
        std::cout << "testUpdate passed" << std::endl;
    } else {
        std::cout << "testUpdate failed" << std::endl;
    }
}

void testMerge() {
    HashMap hm1;
    hm1.fromList({{"key1", "value1"}, {"key2", "value2"}});
    HashMap hm2;
    hm2.fromList({{"key2", "newValue2"}, {"key3", "value3"}});
    hm1.merge(hm2);
    if (hm1.lookup("key1") == "value1" && hm1.lookup("key2") == "value2" && hm1.lookup("key3") == "value3") {
        std::cout << "testMerge passed" << std::endl;
    } else {
        std::cout << "testMerge failed" << std::endl;
    }
}

void testKeys() {
    HashMap hm;
    hm.fromList({{"key1", "value1"}, {"key2", "value2"}});
    auto keys = hm.keys();
    if (keys == std::vector<std::string>{"key1", "key2"}) {
        std::cout << "testKeys passed" << std::endl;
    } else {
        std::cout << "testKeys failed" << std::endl;
    }
}

void testValues() {
    HashMap hm;
    hm.fromList({{"key1", "value1"}, {"key2", "value2"}});
    auto values = hm.values();
    if (values == std::vector<std::string>{"value1", "value2"}) {
        std::cout << "testValues passed" << std::endl;
    } else {
        std::cout << "testValues failed" << std::endl;
    }
}

void testSize() {
    HashMap hm;
    hm.fromList({{"key1", "value1"}, {"key2", "value2"}});
    if (hm.size() == 2) {
        std::cout << "testSize passed" << std::endl;
    } else {
        std::cout << "testSize failed" << std::endl;
    }
}


// New performance test functions
void testInsertPerformance() {
    std::cout << "\nTesting Insert Performance:" << std::endl;
    std::vector<int> sizes = {1000, 10000, 100000};
    
    for (int size : sizes) {
        HashMap hm;
        std::vector<std::pair<std::string, std::string>> data;
        // Prepare data
        for (int i = 0; i < size; i++) {
            data.push_back({randomString(10), randomString(10)});
        }
        
        double time = measureTime([&]() {
            for (const auto& pair : data) {
                hm.insert(pair.first, pair.second);
            }
        });
        
        std::cout << "Insert " << size << " elements: " << time << " seconds" << std::endl;
    }
}

void testLookupPerformance() {
    std::cout << "\nTesting Lookup Performance:" << std::endl;
    HashMap hm;
    std::vector<std::string> keys;
    
    // Insert test data
    for (int i = 0; i < 100000; i++) {
        std::string key = randomString(10);
        keys.push_back(key);
        hm.insert(key, randomString(10));
    }
    
    // Test different lookup patterns
    std::vector<int> lookupCounts = {1000, 10000, 100000};
    for (int count : lookupCounts) {
        double time = measureTime([&]() {
            for (int i = 0; i < count; i++) {
                hm.lookup(keys[i % keys.size()]);
            }
        });
        std::cout << count << " lookups: " << time << " seconds" << std::endl;
    }
}

void testBulkOperations() {
    std::cout << "\nTesting Bulk Operations Performance:" << std::endl;
    std::vector<int> sizes = {1000, 10000, 100000};
    
    for (int size : sizes) {
        std::vector<std::pair<std::string, std::string>> data;
        for (int i = 0; i < size; i++) {
            data.push_back({randomString(10), randomString(10)});
        }
        
        HashMap hm;
        double fromListTime = measureTime([&]() {
            hm.fromList(data);
        });
        
        double toListTime = measureTime([&]() {
            auto list = hm.toList();
        });
        
        std::cout << "Size " << size << ":" << std::endl;
        std::cout << "  fromList: " << fromListTime << " seconds" << std::endl;
        std::cout << "  toList: " << toListTime << " seconds" << std::endl;
    }
}

// Modified main function
int main() {
    // Run original tests
    std::cout << "Running functional tests:" << std::endl;
    testInsert();
    testLookup();
    testDelete();
    testFromList();
    testToList();
    testPurelyFunctionalInsert();
    testPurelyFunctionalDelete();
    testUpdate();
    testMerge();
    testKeys();
    testValues();
    testSize();
    
    // Run performance tests
    testInsertPerformance();
    testLookupPerformance();
    testBulkOperations();
    
    return 0;
}